crud/count.lua

```diff
@@ -0,0 +1,213 @@
+
+local CountError = errors.new_class('CountError', {capture_stack = false})
+
+local COUNT_FUNC_NAME = '_crud.count_on_storage'
+local DEFAULT_COUNT_TO_YIELD = 100
+
+local count = {}
+
+local function count_on_storage(space_name, index_id, conditions, opts)
+    dev_checks('string', 'number', '?table', {
+        scan_value = 'table',
+        tarantool_iter = 'number',
+        count_to_yield = '?number',
+        scan_condition_num = '?number',
+    })
+
+    opts = opts or {}
+
+    local space = box.space[space_name]
+
+    local index = space.index[index_id]
+    if index == nil then
+        return nil, CountError:new("Index with ID %s doesn't exist", index_id)
+    end
+
+    local value = opts.scan_value
+
+    local filter_func, err = filters.gen_func(space, conditions, {
+        tarantool_iter = opts.tarantool_iter,
+        scan_condition_num = opts.scan_condition_num,
+    })
+    if err ~= nil then
+        return nil, CountError:new("Failed to generate tuples filter: %s", err)
+    end
+
+    local tuples_count = 0
+
+    local tuple
+    local gen = index:pairs(value, {iterator = opts.tarantool_iter})
+
+    gen.state, tuple = gen(gen.param, gen.state)
+    while true do
+        if tuple == nil then
+            break
+        end
+
+        local matched, early_exit = filter_func(tuple)
+
+        if matched then
+            tuples_count = tuples_count + 1
+
+            if opts.count_to_yield ~= nil and tuples_count % opts.count_to_yield == 0 then
+                fiber.yield()
+            end
+        elseif early_exit then
+            break
+        end
+
+        gen.state, tuple = gen(gen.param, gen.state)
+    end
+
+    return tuples_count
+end
+
+function count.init()
+    _G._crud.count_on_storage = count_on_storage
+end
+
+--- Calculates the number of tuples by conditions
+--
+-- @function call
+--
+-- @param string space_name
+--  A space name
+--
+-- @param ?table user_conditions
+--  Conditions by which tuples are counted,
+--  default value is empty table
+--
+-- @tparam ?number opts.timeout
+--  Function call timeout, default value is 2
+--
+-- @tparam ?number opts.count_to_yield
+--  Number of tuples processed to yield after,
+--  default value is 100
+--
+-- @tparam ?boolean opts.prefer_replica
+--  Call on replica if it's possible,
+--  default value is `nil`, which works as with `false`
+--
+-- @tparam ?boolean opts.balance
+--  Use replica according to round-robin load balancing
+--  default value is `nil`, which works as with `false`
+--
+-- @tparam ?string opts.mode
+--  vshard call mode, default value is `read`
+--
+-- @return[1] number
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+--
+function count.call(space_name, user_conditions, opts)
+    checks('string', '?table', {
+        timeout = '?number',
+        bucket_id = '?number|cdata',
+        count_to_yield = '?number',
+        prefer_replica = '?boolean',
+        balance = '?boolean',
+        mode = '?string',
```

Comment:
Do we need to check possible values for `mode`? It is a string and can be only `read` or `write`.

Reply:
I have added a check for `mode` option in `common/call.lua`:

```lua
function call.get_vshard_call_name(mode, prefer_replica, balance)
    dev_checks('string', '?boolean', '?boolean')

    if mode ~= 'write' and mode ~= 'read' then
        return nil, CallError:new("Unknown call mode: %s", mode)
    end
...
end

function call.map(func_name, func_args, opts)
    dev_checks('string', '?table', {
        mode = 'string',
        prefer_replica = '?boolean',
        balance = '?boolean',
        timeout = '?number',
        replicasets = '?table',
    })
    opts = opts or {}

    local vshard_call_name, err = call.get_vshard_call_name(opts.mode, opts.prefer_replica, opts.balance)
    if err ~= nil then
        return nil, err
    end
...
end

function call.single(bucket_id, func_name, func_args, opts)
    dev_checks('number', 'string', '?table', {
        mode = 'string',
        prefer_replica = '?boolean',
        balance = '?boolean',
        timeout = '?number',
    })

    local vshard_call_name, err = call.get_vshard_call_name(opts.mode, opts.prefer_replica, opts.balance)
    if err ~= nil then
        return nil, err
    end
...
end
```

I have added some tests in `call_test.lua` as well:

```lua
g.test_map_invalid_mode = function()
    local results, err = g.cluster.main_server.net_box:eval([[
        local call = require('crud.common.call')
        return call.map('say_hi_politely', nil, {mode = 'invalid'})
    ]])

    t.assert_equals(results, nil)
    t.assert_str_contains(err.err, "Unknown call mode: invalid")
end

g.test_single_invalid_mode = function()
    local results, err = g.cluster.main_server.net_box:eval([[
        local call = require('crud.common.call')
        return call.single(1, 'say_hi_politely', nil, {mode = 'invalid'})
    ]])

    t.assert_equals(results, nil)
    t.assert_str_contains(err.err, "Unknown call mode: invalid")
end
```